kvserver/concurrency,tracing: add tracing tags while waiting for locks

andreimatei opened the pull request:
When the evaluation of a request blocks on a lock acquisition, this
patch annotates the tracing span representing the evaluation with
tags about the lock's key, holder, time when the waiting started, and
number of locks that the evaluation waited on previously.

To do this, the patch adds support in the tracing library for "lazy
tags": tags that are only stringified on demand, when a span's recording
is collected. This is useful if the stringification is expensive (so we
only want to pay for it if someone actually looks at the span), and also
for structured tags whose state can change over time. A lazy tag can
implement one of two interfaces: Stringer or LazyTag. The latter allows
the tag to be represented as a set of key-value pairs in a recording.

The point of the new contention tags is to be featured prominently in
the visualizer for the active spans registry that I'm working on.
Looking at queries/requests experiencing contention is likely to be a
very common scenario.

This patch spruces up the contentionEventHelper in the lock table, to
maintain the state of a waiter as a lazy tag.

Release note: None
Release justification: Low risk, high observability value.

andreimatei committed:
tracing: small cleanup

An inner function is no longer serving any purpose.

Release note: None

andreimatei committed:
kvserver/concurrency,tracing: add tracing tags while waiting for locks

When the evaluation of a request blocks on a lock acquisition, this
patch annotates the tracing span representing the evaluation with
tags about the lock's key, holder, time when the waiting started, and
number of locks that the evaluation waited on previously.

To do this, the patch adds support in the tracing library for "lazy
tags": tags that are only stringified on demand, when a span's recording
is collected. This is useful if the stringification is expensive (so we
only want to pay for it if someone actually looks at the span), and also
for structured tags whose state can change over time. A lazy tag can
implement one of two interfaces: Stringer or LazyTag. The latter allows
the tag to be represented as a set of key-value pairs in a recording.

The point of the new contention tags is to be featured prominently in
the visualizer for the active spans registry that I'm working on.
Looking at queries/requests experiencing contention is likely to be a
very common scenario.

This patch spruces up the contentionEventHelper in the lock table, to
maintain the state of a waiter as a lazy tag.

Release note: None
Release justification: Low risk, high observability value.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74878)
<!-- Reviewable:end -->

andreimatei commented:
@nvanbenschoten remind me pls - if a request is looking to acquire locks itself, then the `newStateC` in the `lockTableWaiterImpl` is only ever signal with one lock to wait on; when that first lock becomes available, the request loops around in `executeBatchWithConcurrencyRetries`, takes the lock, and starts waiting on the next one, right? As opposed to a request that doesn't want to acquire anything (so it just wants for the conflicting locks to go away), for which the `newStateC` can be signaled with multiple locks, right? If so, who controls this distinction - signaling `newStateC` multiple times or not?

tbg commented on pkg/kv/kvserver/concurrency/lock_table_waiter.go:
```diff
@@ -901,17 +924,65 @@ func (h *contentionEventHelper) emitAndInit(s waitingState) {
+// 1m01s
+// 1h05m01s
+// 1d02h05m
+func formatDuration(d time.Duration) string {
```
https://github.com/cockroachdb/cockroach/blob/6798516f9d5f7b06d3249bb113dc34309c9b102b/pkg/util/humanizeutil/duration.go#L20-L29?

tbg commented on pkg/kv/kvserver/concurrency/lock_table_waiter.go:
```diff
@@ -901,17 +924,65 @@ func (h *contentionEventHelper) emitAndInit(s waitingState) {
 			h.tBegin = timeutil.Now()
 		}
-	case waitElsewhere, waitQueueMaxLengthExceeded, doneWaiting:
+		h.sp.SetStatusTag(tagWaitKey, attribute.StringValue(s.key.String()))
```
The analog to https://github.com/cockroachdb/cockroach/pull/71622 could also be helpful here. I know, you're trying to get us to a place where we have no need for such stop-gaps, but still, it might get us places for the foreseeable future. I would certainly appreciate having this in the debug.zip (where the labels [end up](https://github.com/cockroachdb/cockroach/pull/74070))

tbg commented on pkg/util/tracing/crdbspan.go:
```diff
@@ -394,6 +394,20 @@ func (s *crdbSpan) setTagLocked(key string, value attribute.Value) {
+	defer s.mu.Unlock()
+	k := attribute.Key(key)
+	for i := range s.mu.tags {
+		if s.mu.tags[i].Key == k {
```
Should store them separately, otherwise you can remove regular tags, no?

tbg commented:
Can you have someone from @cockroachdb/obs-inf-prs on this PR since they own tracing? I think there are probably some finer points to discuss.

The overall idea LGTM.

dhartunian reviewed (commented):
> Reviewed 1 of 1 files at r1, 1 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74878)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @andreimatei, @dhartunian, and @nvanbenschoten)

___
*[-- commits, line 14 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mti-kr7BGfitXv8-yy7):*
nit: waited
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 855 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mti068XDOgnZr56zKqt:-Mti068XDOgnZr56zKqu:b-dl0y42) ([raw file](https://github.com/cockroachdb/cockroach/blob/b32e587c4113d56f9fee809f43148b4c58d09cc5/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L855)):*
> ```Go
> // A contentionEventHelper should be created when a request starts waiting in
> // the lock table (on what might turn out to be multiple locks). close() should
> // be called when the waiting is over.
> ```

Just curious in terms of API design, if this should mirror a lock with a `start` method to match `close`. If I create an instance of this struct without a start time, it seems like that would break things.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 862 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mti1hyyB_y3x4CyhRww:-Mti1hyyB_y3x4CyhRwx:beecbpb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b32e587c4113d56f9fee809f43148b4c58d09cc5/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L862)):*
> ```Go
> 	// contentionStart marks the start of the contention tracked by this
> 	// contentionEventHelper.
> 	contentionStart time.Time
> ```

I'm a bit confused about the difference between `tBegin` below and `contentionStart`. Can you clarify the difference between "start of the contention" and "start of the contention event"? Also, is there a meaning to using both `Begin` and `Start`?
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 944 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mti0jng5z9bwuaoylw-:-Mti0jng5z9bwuaoylw0:b-kwfhtu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b32e587c4113d56f9fee809f43148b4c58d09cc5/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L944)):*
> ```Go
> 
> func (h *contentionEventHelper) close() {
> 	if h.closed || h.sp == nil {
> ```

Do we ever set closed to `true`?
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 948 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MtiOn_F3WPtn8YxigPT:-MtiOn_F3WPtn8YxigPU:bj3z7dx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b32e587c4113d56f9fee809f43148b4c58d09cc5/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L948)):*
> ```Go
> 	}
> 	// Clear most tags.
> 	for _, tag := range []string{"lock_wait_key", "lock_wait_start", "lock_holder_txn"} {
> ```

Just checking my understanding here: These tags will only be visible while the span is actually waiting on the locks, right? So you have to "catch" the span while it's stuck in order to see these tags?


<!-- Sent from Reviewable.io -->

andreimatei committed:
kvserver/concurrency: account for waitElsewhere time in contention events

Before this patch, no contention event was produced when requests were
waiting for locks by pushing the txn that currently has a conflicting
lock (or a reservation on a conflicting lock). This corresponds to the
waitElsewhere signal from the lock table. (waitElsewhere is not the
common case; the common case is for a request to wait in the lock table
(through the waitFor signal). waitElsewhere happens instead of waitFor
when the lock table is under memory pressure and sheds its state.)

This patch makes the waitElsewhere state be handled like the other
waiting states for the purposes of generating contention events.

This patch also adds a data-driven test for the waitElsewhere
functionality.

Release note: None

andreimatei committed:
util/tracing: remove stale comment about tags

The comment was claiming that a span's tag are only maintained if the
span is recording, but that hasn't been the case in a while.

Release note: None

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74878)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @dhartunian, @nvanbenschoten, and @tbg)

___
*[-- commits, line 14 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mti-kr7BGfitXv8-yy7):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: waited
</blockquote></details>

done
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 855 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mti068XDOgnZr56zKqt:-MuO-Swr28nByI1uJ4c6:b-7scxk9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b32e587c4113d56f9fee809f43148b4c58d09cc5/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L855)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Just curious in terms of API design, if this should mirror a lock with a `start` method to match `close`. If I create an instance of this struct without a start time, it seems like that would break things.
</blockquote></details>

I've given it a constructor.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 862 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mti1hyyB_y3x4CyhRww:-MuTKfYj3644L0luwx5q:b-o341uf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b32e587c4113d56f9fee809f43148b4c58d09cc5/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L862)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I'm a bit confused about the difference between `tBegin` below and `contentionStart`. Can you clarify the difference between "start of the contention" and "start of the contention event"? Also, is there a meaning to using both `Begin` and `Start`?
</blockquote></details>

please see now; I've reworked this.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 927 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MtPswYNx5PEBbfd3gS--r2-927:-MuTKOGj5EWagJ8SGFBU:b4teumt) ([raw file](https://github.com/cockroachdb/cockroach/blob/b32e587c4113d56f9fee809f43148b4c58d09cc5/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L927)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

The analog to https://github.com/cockroachdb/cockroach/pull/71622 could also be helpful here. I know, you're trying to get us to a place where we have no need for such stop-gaps, but still, it might get us places for the foreseeable future. I would certainly appreciate having this in the debug.zip (where the labels [end up](https://github.com/cockroachdb/cockroach/pull/74070))
</blockquote></details>

What if I promise to put trace registry dumps in the debug.zip? :)
Regardless, I'd rather not put that label in this patch. Let's discuss it separately.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 944 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mti0jng5z9bwuaoylw-:-MuTJ8E71ATmJBUSTKMb:b-t29ljz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b32e587c4113d56f9fee809f43148b4c58d09cc5/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L944)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Do we ever set closed to `true`?
</blockquote></details>

we do now :)
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 948 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MtiOn_F3WPtn8YxigPT:-MuTJACs0XCf92l9-GHn:b-kuwer0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b32e587c4113d56f9fee809f43148b4c58d09cc5/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L948)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Just checking my understanding here: These tags will only be visible while the span is actually waiting on the locks, right? So you have to "catch" the span while it's stuck in order to see these tags?
</blockquote></details>

I've reworked this. Now, there is technically a single tag that stays with the span forever, and that span renders differently depending on whether the span is currently blocked versus when it's no longer blocked. In the latter case, it just has some aggregate info about the total amount of time spent waiting on locks.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 963 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MtPswYNx5PEBbfd3gS--r2-963:-MuTJr1q5Pn-QL28CIgn:b-5v1bv6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b32e587c4113d56f9fee809f43148b4c58d09cc5/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L963)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

https://github.com/cockroachdb/cockroach/blob/6798516f9d5f7b06d3249bb113dc34309c9b102b/pkg/util/humanizeutil/duration.go#L20-L29?
</blockquote></details>

thanks, used that one
___
*[pkg/util/tracing/crdbspan.go, line 403 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MtPswYNx5PEBbfd3gS0-r2-403:-MuTJwTD-D5heU7vxH82:b-ve4a1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/b32e587c4113d56f9fee809f43148b4c58d09cc5/pkg/util/tracing/crdbspan.go#L403)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Should store them separately, otherwise you can remove regular tags, no?
</blockquote></details>

yeah... I purposefully didn't go through that trouble. When no Otel tracer is attached, removing regular tags works just as well as removing "status tags", if one does it, and I think that's fine.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
@alextalks can I please ask you for a review for the kv parts, and @dhartunian for the tracing library changes?
And @tbg, @nvanbenschoten, for whatever you feel like :).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-:-MuTKpiG5xfzgqEls33H:bwbm1e6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @dhartunian, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->

AlexTalks reviewed (changes requested):
> Reviewed 9 of 9 files at r7, 1 of 1 files at r8, 3 of 3 files at r9, 5 of 7 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74878)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @andreimatei, @dhartunian, @nvanbenschoten, and @tbg)

___
*[-- commits, line 22 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MutSsNn7S3Z1bihQNyp):*
The changes make sense but this commit message may be a little confusing - as you mentioned, `waitElsewhere` is basically the same as `waitFor`, it's just that the lock table is under memory pressure (the number of locks tracked has exceeded configured limits for this replica).  It might be that a slight rewording is necessary to make this clear here.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 896 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MutXIk4-L5kODNkfCh2:-MutXIk4-L5kODNkfCh3:b-s63p6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L896)):*
> ```Go
> 		closed bool
> 
> 		// lockWait represents the time the span spent waiting on locks. If closed
> ```

the **cumulative** time the **request* spent waiting on locks
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 906 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MutZ9dcELOXJ13Bp3mD:-MutZ9dcELOXJ13Bp3mE:b-xrbd9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L906)):*
> ```Go
> 		// closed or if the contentionEventTracer has not yet been notified of the
> 		// first lock to wait on.
> 		curState *waitingState
> ```

I'm not sure it makes sense to additionally track this state on the `contentionEventHelper` when it seems state change tracking logic could be accomplished using local variables in `WaitOn(..)`, no?
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 931 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mut_sRY1P3wb5yaAClx:-Mut_sRY1P3wb5yaACly:b9s0f7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L931)):*
> ```Go
> 	t, ok := sp.GetLazyTag(tagContentionHelper)
> 	if ok {
> 		h = t.(*contentionEventTracer)
> ```

It may be worth discussing if having a per-span `contentionEventTracer` is the best way to do this, or if it would make sense to capture discrete events by contended txn/key and then be able to look at the individual contentions (and their durations).

You could potentially still have a per-span `contentionEventTracer` but collect each individual event in it (along with duration) so in `ExpandToRecordingTags(..)` you could still collect the cumulative duration of contentions (or, of course, render it a different way).


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> Reviewed 9 of 9 files at r7, 1 of 1 files at r8, 3 of 3 files at r9, 2 of 7 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74878)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @andreimatei, @dhartunian, @nvanbenschoten, and @tbg)

___
*[-- commits, line 38 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mv15lYT-JjWeUjbUoc0):*
I want to check my understanding of the data we're collecting and where/when it will get surfaced:

* When the span is waiting on a lock and we capture it in-flight, we'll get a snapshot of the `contentionEventTracer` state within via the `ExpandingTag` mechanism.
* If the span waited on locks *in the past* these each produce `ContentionEvent`s via the `emit()` call that add structured data to the span.

Does the `contentionEventTracer` leave data in its wake besides the structured `ContentionEvent`s?
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 846 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mv0jOGN8lnPce6nBKMg:-Mv0jOGN8lnPce6nBKMh:b-sku83g) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L846)):*
> ```Go
> // tagContentionHelper is the tracing span tag that the *contentionEventHelper
> // lives under.
> const tagContentionHelper = "contention_helper"
> ```

this should be `content_tracer` now?

Also a bunch of other cases of  `helper` being used below that should be changed to `tracer`.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 865 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mv16x9J1E6M6GZMle4I:-Mv16x9J1E6M6GZMle4J:bugzdf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L865)):*
> ```Go
> // waiting on, if any, is not
> // counted.
> const tagNumLocks = "lock_num"
> ```

nit: can we name this to match the string?
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 870 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mv0De1u8G3vKUIBz8Kt:-Mv0De1vC8RRi4sjRJEh:batgrs3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L870)):*
> ```Go
> // waited on locks. If the span is currently waiting on a lock, the time it has
> // already waited on this lock is not reflected.
> const tagWaited = "lock_wait"
> ```

checking my understanding: this is recorded after the lock is released so that's why "current" wait time is not factored in?
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 935 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mv0k-xj1CWgL7iIuVj2:-Mv0k-xj1CWgL7iIuVj3:b-1i9jbb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L935)):*
> ```Go
> 		defer h.mu.Unlock()
> 		if !h.mu.closed {
> 			log.Fatalf(ctx, "unexpectedly found non-closed contentionEventTracer")
> ```

does this really warrant `Fatal`? Or is this more meant for catching errors during test?
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 1028 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mv0mPCeBDNk642vya9G:-Mv0mPCeBDNk642vya9H:b-7t8o14) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L1028)):*
> ```Go
> 		// NB: we're guaranteed to have `s.{txn,key}` populated here.
> 		differentLock := false
> 		var firstLock bool
> ```

random q: why initialize these variables differently?
___
*[pkg/util/tracing/span_inner.go, line 168 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mv18yO4B21AIEcMtvWl:-Mv18yO4B21AIEcMtvWm:b-ysi67f) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/util/tracing/span_inner.go#L168)):*
> ```Go
> }
> 
> func (s *spanInner) SetLazyTag(key string, value interface{}, statusTag bool) *spanInner {
> ```

The API here expands quite a bit, do you think it's worth consolidating a bit? can "lazy" be a bool like `statusTag`?


<!-- Sent from Reviewable.io -->

andreimatei commented:
Sorry to let this sit; please give it another look.
I've simplified the changes to the span api.

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Much easier to follow. Thanks for making these changes. Just some nits re: unused methods being added.

> Reviewed 9 of 9 files at r11, 1 of 1 files at r12, 3 of 3 files at r13, 3 of 7 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-:-Mxe_GWy-bhNdZLqskxF:b-e9s2d6)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @dhartunian, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 1004 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mxe_3oq0tSbdc6bo8y6:-Mxe_3oq0tSbdc6bo8y7:baz09hf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L1004)):*
> ```Go
> // notify processes an event from the lock table.
> // compares the waitingState's active txn (if any) against the current
> // ContentionEvent (if any). If the they match, we are continuing to handle the
> ```

nit: `the they`
___
*[pkg/util/tracing/crdbspan.go, line 536 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MxeYCXP5TsDiWKJ_MYC:-MxeYCXP5TsDiWKJ_MYD:b-k0rhnh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a70534cbe215c187f8d9cc0789363f9d31b66bdd/pkg/util/tracing/crdbspan.go#L536)):*
> ```Go
> 
> // clearTag removes a tag, if it exists.
> func (s *crdbSpan) clearTag(key string) {
> ```

This seems unused. Only called by `spanInner.ClearStatusTag` which is also unused.
___
*[pkg/util/tracing/span_inner.go, line 154 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MxeYkU75j0Y4JvpJhuq:-MxeYkU75j0Y4JvpJhur:b-z0kgsq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a70534cbe215c187f8d9cc0789363f9d31b66bdd/pkg/util/tracing/span_inner.go#L154)):*
> ```Go
> }
> 
> func (s *spanInner) SetTag(key string, value attribute.Value, statusTag bool) *spanInner {
> ```

I don't see a usage that sets `statusTag` to `true`. Am I missing something?
___
*[pkg/util/tracing/span_inner.go, line 196 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MxeYvIW3qD2_-SjjpuZ:-MxeYvIW3qD2_-Sjjpu_:bntsdp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a70534cbe215c187f8d9cc0789363f9d31b66bdd/pkg/util/tracing/span_inner.go#L196)):*
> ```Go
> }
> 
> func (s *spanInner) ClearStatusTag(key string) *spanInner {
> ```

Unused. Same as `clearTag()`.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74878)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @AlexTalks, @dhartunian, @nvanbenschoten, and @tbg)

___
*[-- commits, line 22 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MutSsNn7S3Z1bihQNyp):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

The changes make sense but this commit message may be a little confusing - as you mentioned, `waitElsewhere` is basically the same as `waitFor`, it's just that the lock table is under memory pressure (the number of locks tracked has exceeded configured limits for this replica).  It might be that a slight rewording is necessary to make this clear here.
</blockquote></details>

I added a few words, please see now
___
*[-- commits, line 38 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mv15lYT-JjWeUjbUoc0):*
> When the span is waiting on a lock and we capture it in-flight, we'll get a snapshot of the contentionEventTracer state within via the ExpandingTag mechanism.

Right. Except that you'll get some data from the `ExpandingTag` even if the span's recording is captured after the span finished waiting; see below.

> If the span waited on locks in the past these each produce ContentionEvents via the emit() call that add structured data to the span.

Right. But note that, at least currently, the structured events are no-ops on non-recording spans, just like non-structured events.
Besides the structured events, the `contentionEventTracer` also counts the number of locks and the total time waited. See below.

> Does the contentionEventTracer leave data in its wake besides the structured ContentionEvents?

It does. Or, rather, there's no "in its wake"; once created, the `contentionEventTracer` stays attached to the span as a `LazyTag` until `span.Finish()`. If/when the span's recording is collected, it will include the number of locks and the total time waited. If the recording is collected while the span is waiting on a lock, we get extra info about that lock.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 846 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mv0jOGN8lnPce6nBKMg:-MxBouDa-Z42VWQAr2u1:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L846)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

this should be `content_tracer` now?

Also a bunch of other cases of  `helper` being used below that should be changed to `tracer`.
</blockquote></details>

done
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 865 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mv16x9J1E6M6GZMle4I:-MxBoyozESVQBLx21hSJ:b-d2c26y) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L865)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: can we name this to match the string?
</blockquote></details>

I'd prefer not to, unless you feel strongly. I think it's useful for all these consts to have a common prefix, so that code completion groups them. When rendered as tags in a span, this consideration no longer applies. 
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 870 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mv0De1u8G3vKUIBz8Kt:-MxBtlsp97pDvOxFw0jA:b-yiti04) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L870)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

checking my understanding: this is recorded after the lock is released so that's why "current" wait time is not factored in?
</blockquote></details>

that was the case, but I've now changed it to include the current current lock (if any)
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 896 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MutXIk4-L5kODNkfCh2:-MxC3D1T855DsZh8xN_A:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L896)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

the **cumulative** time the **request* spent waiting on locks
</blockquote></details>

done
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 906 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MutZ9dcELOXJ13Bp3mD:-MxC7q_kDHr6BAzA7Gsj:bu6hn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L906)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

I'm not sure it makes sense to additionally track this state on the `contentionEventHelper` when it seems state change tracking logic could be accomplished using local variables in `WaitOn(..)`, no?
</blockquote></details>

well, `curState` is used in `Render()`, which is called async - there's no opportunity for the `lockTableWaiterImpl` to provide data in that case.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 931 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mut_sRY1P3wb5yaAClx:-MxC43l35AVSp0z2HwNi:b-hechp9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L931)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

It may be worth discussing if having a per-span `contentionEventTracer` is the best way to do this, or if it would make sense to capture discrete events by contended txn/key and then be able to look at the individual contentions (and their durations).

You could potentially still have a per-span `contentionEventTracer` but collect each individual event in it (along with duration) so in `ExpandToRecordingTags(..)` you could still collect the cumulative duration of contentions (or, of course, render it a different way).
</blockquote></details>

The point of only storing aggregate information (in addition to extra info about the current lock being waited on, if any), is so that all this tracing takes a fixed memory budget when recording is not enabled. When recording is enabled, we get the structured events after each lock.
You don't buy?
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 935 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mv0k-xj1CWgL7iIuVj2:-MxC4ewo2MTuvtJtj5KT:b-7oh1nb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L935)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

does this really warrant `Fatal`? Or is this more meant for catching errors during test?
</blockquote></details>

I personally am in the crash-happy camp for bugs, on the argument that it generally leads to better and tighter code. If somehow `getOrCreateContentionEventHelper` were called with a span with an open `contentionEventHelper`, then either we forgot to `close()` it (in which case we're been counting bogus wait time), or we're somehow using the span for two concurrent wait table interactions, in which case they would step over each other. Both are a form of data corruption. Unless there's good reason to relax this, I'd rather generally assert that my data is sane.
You object?
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 1004 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mxe_3oq0tSbdc6bo8y6:-Mxl81iK3jc0qWJ2gO8X:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L1004)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: `the they`
</blockquote></details>

done
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 1028 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mv0mPCeBDNk642vya9G:-MxC6L8q9hR7c1gsIOT2:bo0hxs5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L1028)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

random q: why initialize these variables differently?
</blockquote></details>

changed now
___
*[pkg/util/tracing/crdbspan.go, line 536 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MxeYCXP5TsDiWKJ_MYC:-Mxl8BX521GyAIVDTbzo:b-u3i7m0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a70534cbe215c187f8d9cc0789363f9d31b66bdd/pkg/util/tracing/crdbspan.go#L536)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This seems unused. Only called by `spanInner.ClearStatusTag` which is also unused.
</blockquote></details>

removed, thanks
___
*[pkg/util/tracing/span_inner.go, line 168 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mv18yO4B21AIEcMtvWl:-MxC6p5m0qtogO1_F0q7:b-zgfmrt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/util/tracing/span_inner.go#L168)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

The API here expands quite a bit, do you think it's worth consolidating a bit? can "lazy" be a bool like `statusTag`?
</blockquote></details>

please see now; I've simplified things
___
*[pkg/util/tracing/span_inner.go, line 154 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MxeYkU75j0Y4JvpJhuq:-Mxl8QSrDrNbBCCdFmvF:b-klwakh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a70534cbe215c187f8d9cc0789363f9d31b66bdd/pkg/util/tracing/span_inner.go#L154)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I don't see a usage that sets `statusTag` to `true`. Am I missing something?
</blockquote></details>

this was detritus; removed
___
*[pkg/util/tracing/span_inner.go, line 196 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MxeYvIW3qD2_-SjjpuZ:-Mxl8VIT4zdVrzotdfb3:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/a70534cbe215c187f8d9cc0789363f9d31b66bdd/pkg/util/tracing/span_inner.go#L196)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Unused. Same as `clearTag()`.
</blockquote></details>

removed


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (changes requested):
> Reviewed 1 of 7 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74878)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @dhartunian, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 931 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mut_sRY1P3wb5yaAClx:-Mxp_75KBfCDgi59N4As:bq034zg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L931)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

The point of only storing aggregate information (in addition to extra info about the current lock being waited on, if any), is so that all this tracing takes a fixed memory budget when recording is not enabled. When recording is enabled, we get the structured events after each lock.
You don't buy?
</blockquote></details>

I think that makes sense. I'm not entirely sure how to reason about what our memory budget is here, but I understand the thought process.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 882 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MxpYhwR5oVfgEzjh8aN:-MxpYhwR5oVfgEzjh8aO:b68ku9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbf2e8960ae0c49204b341b18195c261ded467de/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L882)):*
> ```Go
> // over.
> //
> // contentionEventTracer is thread-safe because ExpandToRecordingTags can be
> ```

`s/ExpandToRecordingTags/Render`
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 896 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MxpUjKjAMziEO9ZoNbl:-MxpUjKjAMziEO9ZoNbm:bpazr8m) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbf2e8960ae0c49204b341b18195c261ded467de/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L896)):*
> ```Go
> 		// lockWait accumulates time waited for locks when the contentionEventTracer
> 		// is closed.
> 		lockWait time.Duration
> ```

Very useful!

It could potentially be even more useful to make this available to `lockTableGuardImpl` so that it could be used in metrics, though that may be out of scope here.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 913 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MxpV0Ie7wqKjToHMc6X:-MxpV0Ie7wqKjToHMc6Y:bu1mxjh) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbf2e8960ae0c49204b341b18195c261ded467de/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L913)):*
> ```Go
> 		// not necessarily the last update - if two successive states refer to the
> 		// same	lock, curLockStart is not updated.
> 		curLockStart time.Time
> ```

This may duplicate `lockTableGuardImpl.mu.requestWaitBegin`, which I had initially included in `waitingState` (which would have made it available here).  Perhaps this should be refactored to avoid duplication?  @nvanbenschoten any thoughts?

(Also, that said, I don't think that allows us to account for `waitElsewhere`)
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 940 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MxpS4k4BoTKj__WEMnn:-MxpS4k4BoTKj__WEMno:b-w2ebbq) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbf2e8960ae0c49204b341b18195c261ded467de/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L940)):*
> ```Go
> 		}
> 		h.mu.closed = false
> 		h.mu.waitStart = timeutil.Now()
> ```

Nit: If it does make sense to continue tracking this here, the `lockTableImpl` is created with a `timeutil.TimeSource` that helps with testing - it would be good to pass that in to `lockTableWaiterImpl`, and use that instead of `timeutil.Now()`.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 947 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MxpT4ngBN2p6UWnhe1D:-MxpT4ngBN2p6UWnhe1E:bigi4ov) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbf2e8960ae0c49204b341b18195c261ded467de/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L947)):*
> ```Go
> 		onEvent: onEvent,
> 	}
> 	h.mu.waitStart = timeutil.Now()
> ```

Nit: Same as above.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74878)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @dhartunian, @nvanbenschoten, and @tbg)

___
*[-- commits, line 38 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mv15lYT-JjWeUjbUoc0):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

> When the span is waiting on a lock and we capture it in-flight, we'll get a snapshot of the contentionEventTracer state within via the ExpandingTag mechanism.

Right. Except that you'll get some data from the `ExpandingTag` even if the span's recording is captured after the span finished waiting; see below.

> If the span waited on locks in the past these each produce ContentionEvents via the emit() call that add structured data to the span.

Right. But note that, at least currently, the structured events are no-ops on non-recording spans, just like non-structured events.
Besides the structured events, the `contentionEventTracer` also counts the number of locks and the total time waited. See below.

> Does the contentionEventTracer leave data in its wake besides the structured ContentionEvents?

It does. Or, rather, there's no "in its wake"; once created, the `contentionEventTracer` stays attached to the span as a `LazyTag` until `span.Finish()`. If/when the span's recording is collected, it will include the number of locks and the total time waited. If the recording is collected while the span is waiting on a lock, we get extra info about that lock.
</blockquote></details>

Ack. Thankyou!
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 935 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-Mv0k-xj1CWgL7iIuVj2:-MxplAdP9qlAIWQisbtY:b-z8edv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f204bd297d7800a8318c391fdfab386e89a5d625/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L935)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I personally am in the crash-happy camp for bugs, on the argument that it generally leads to better and tighter code. If somehow `getOrCreateContentionEventHelper` were called with a span with an open `contentionEventHelper`, then either we forgot to `close()` it (in which case we're been counting bogus wait time), or we're somehow using the span for two concurrent wait table interactions, in which case they would step over each other. Both are a form of data corruption. Unless there's good reason to relax this, I'd rather generally assert that my data is sane.
You object?
</blockquote></details>

Makes sense, thx for clarifying :thumbsup: 


<!-- Sent from Reviewable.io -->

andreimatei committed:
concurrency: rename lockTableGuard.requestWaitBegin

Rename to curLockWaitStart to emphasize that it is set once per lock,
not once per request or once per wait state.

Release note: None

andreimatei committed:
concurrency: remove superfluous setting of curLockWaitStart

waitQueueMaxLengthExceeded is not really a wait state, so setting
curLockWaitStart for it doesn't make much sense. Other final states like
also don't update curLockWaitStart.

Release note: None

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74878)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @AlexTalks, @dhartunian, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 882 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MxpYhwR5oVfgEzjh8aN:-MxpfkOJ0R-xBREtiHX-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbf2e8960ae0c49204b341b18195c261ded467de/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L882)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

`s/ExpandToRecordingTags/Render`
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 896 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MxpUjKjAMziEO9ZoNbl:-Mxq9DJ-5NdlsSIC4jCA:b-z50kb8) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbf2e8960ae0c49204b341b18195c261ded467de/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L896)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Very useful!

It could potentially be even more useful to make this available to `lockTableGuardImpl` so that it could be used in metrics, though that may be out of scope here.
</blockquote></details>

yeah... I'll leave this for next time.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 913 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MxpV0Ie7wqKjToHMc6X:-Mxq9GYZAlBVBL3VpfHF:brmzd2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbf2e8960ae0c49204b341b18195c261ded467de/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L913)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

This may duplicate `lockTableGuardImpl.mu.requestWaitBegin`, which I had initially included in `waitingState` (which would have made it available here).  Perhaps this should be refactored to avoid duplication?  @nvanbenschoten any thoughts?

(Also, that said, I don't think that allows us to account for `waitElsewhere`)
</blockquote></details>

as discussed offline, I've plumbed `requestWaitBegin` through the `waitState` over from the `lockTableGuard` to the `contentionEventTracer`.

I've also prepended a commit to rename `requestWaitBegin -> curLockWaitStart` if you don't mind, to reflect the understanding that the field is scoped to a lock.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 940 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MxpS4k4BoTKj__WEMnn:-Mxq9XjR5skO4pDSGGjS:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbf2e8960ae0c49204b341b18195c261ded467de/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L940)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Nit: If it does make sense to continue tracking this here, the `lockTableImpl` is created with a `timeutil.TimeSource` that helps with testing - it would be good to pass that in to `lockTableWaiterImpl`, and use that instead of `timeutil.Now()`.
</blockquote></details>

removed
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 947 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MxpT4ngBN2p6UWnhe1D:-MxqF_vX7yf43-H6MwAo:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbf2e8960ae0c49204b341b18195c261ded467de/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L947)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Nit: Same as above.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

AlexTalks reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 9 of 9 files at r19, 1 of 1 files at r20.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-:-Mxr1nfa5zzggKjRlRZY:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @AlexTalks, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 9 of 9 files at r19, 1 of 1 files at r20, 3 of 3 files at r21, 1 of 1 files at r22, 1 of 1 files at r23, 9 of 9 files at r24, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74878)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei and @tbg)

___
*[pkg/kv/kvserver/concurrency/concurrency_control.go, line 719 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-My8iT7s3ZvLv8v17vij:-My8iT7s3ZvLv8v17vik:b1mrn2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d9ba98fceed0dd6f87cfaa31b5948cedfc03e73/pkg/kv/kvserver/concurrency/concurrency_control.go#L719)):*
> ```Go
> 	Metrics() LockTableMetrics
> 
> 	// Clock returns the lockTable's time source.
> ```

Instead of adding to this interface, let's give `lockTableWaiterImpl` a new `contentionEventClock` field that gets set to the same time source on construction.
___
*[pkg/kv/kvserver/concurrency/lock_table.go, line 96 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-My8ilMb0rJcwxWnfjvu:-My8ilMc9vqGZyUb-6SP:bt99yuj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d9ba98fceed0dd6f87cfaa31b5948cedfc03e73/pkg/kv/kvserver/concurrency/lock_table.go#L96)):*
> ```Go
> 	queuedWriters int               // how many writers are waiting?
> 	queuedReaders int               // how many readers are waiting?
> 	// lockWaitStart represents the timestamp when the request started waiting on
> ```

nit: move below `guardAccess`. Otherwise though, I'm glad we decided to pass this information through the `waitingState`.
___
*[pkg/kv/kvserver/concurrency/lock_table.go, line 1562 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-My8jGUM3QetH_toXq1r:-My8jGUM3QetH_toXq1s:bssnntk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d9ba98fceed0dd6f87cfaa31b5948cedfc03e73/pkg/kv/kvserver/concurrency/lock_table.go#L1562)):*
> ```Go
> 	g.mu.Lock()
> 	lockWaitStart := g.mu.curLockWaitStart
> 	g.mu.Unlock()
> ```

Can we move this down to the existing critical section blow? It looks like. we only need it right after `g.mu.curLockWaitStart = timeProvider.Now()`.
___
*[pkg/kv/kvserver/concurrency/lock_table.go, line 2046 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-My8jhEK-T9u94mqetQC:-My8jhEK-T9u94mqetQD:bofz9ac) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d9ba98fceed0dd6f87cfaa31b5948cedfc03e73/pkg/kv/kvserver/concurrency/lock_table.go#L2046)):*
> ```Go
> 		g.mu.Lock()
> 		g.mu.state = waitState
> 		g.mu.state.lockWaitStart = g.mu.curLockWaitStart
> ```

Assigning to `g.mu.state.lockWaitStart` is nice. It prevents lockWaitStart times from accidentally ending up in the wrong guard's state. We should do that everywhere else as well.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 887 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-My8ftNb44vSb3E_2PvY:-My8ftNb44vSb3E_2PvZ:bphzp5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/81efd64752251c5fe126c738c5b64c465ac9e600/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L887)):*
> ```Go
> 	// Otherwise,
> 	switch s.kind {
> 	case waitFor, waitForDistinguished, waitSelf, waitElsewhere:
> ```

Should we write a test for this? Similar to `pkg/kv/kvserver/concurrency/testdata/concurrency_manager/wait_self`, which was added in https://github.com/cockroachdb/cockroach/commit/dc689ae9bdef495533c5aef2cc294485186201d3.
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 908 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-My8k9k22YiSxeTP8hVw:-My8k9k22YiSxeTP8hVx:b-v6x8jh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d9ba98fceed0dd6f87cfaa31b5948cedfc03e73/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L908)):*
> ```Go
> 		// closed or if the contentionEventTracer has not yet been notified of the
> 		// first lock to wait on.
> 		curState *waitingState
> ```

Does this need to be a pointer? I assume it causes the argument to `notify` to escape to the heap and raises (unfounded) aliasing concerns.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74878)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @tbg)

___
*[pkg/kv/kvserver/concurrency/concurrency_control.go, line 719 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-My8iT7s3ZvLv8v17vij:-My9Ac4D07BMipLSa88J:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d9ba98fceed0dd6f87cfaa31b5948cedfc03e73/pkg/kv/kvserver/concurrency/concurrency_control.go#L719)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Instead of adding to this interface, let's give `lockTableWaiterImpl` a new `contentionEventClock` field that gets set to the same time source on construction.
</blockquote></details>

done
___
*[pkg/kv/kvserver/concurrency/lock_table.go, line 96 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-My8ilMb0rJcwxWnfjvu:-My8qyk3CPEmyxRL6tQC:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d9ba98fceed0dd6f87cfaa31b5948cedfc03e73/pkg/kv/kvserver/concurrency/lock_table.go#L96)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: move below `guardAccess`. Otherwise though, I'm glad we decided to pass this information through the `waitingState`.
</blockquote></details>

done
___
*[pkg/kv/kvserver/concurrency/lock_table.go, line 1562 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-My8jGUM3QetH_toXq1r:-My8sa50AFN39V0i2UAT:b-2h81s0) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d9ba98fceed0dd6f87cfaa31b5948cedfc03e73/pkg/kv/kvserver/concurrency/lock_table.go#L1562)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we move this down to the existing critical section blow? It looks like. we only need it right after `g.mu.curLockWaitStart = timeProvider.Now()`.
</blockquote></details>

done. Please see now,
___
*[pkg/kv/kvserver/concurrency/lock_table.go, line 2046 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-My8jhEK-T9u94mqetQC:-My91P__DRn48HrGTtQL:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d9ba98fceed0dd6f87cfaa31b5948cedfc03e73/pkg/kv/kvserver/concurrency/lock_table.go#L2046)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Assigning to `g.mu.state.lockWaitStart` is nice. It prevents lockWaitStart times from accidentally ending up in the wrong guard's state. We should do that everywhere else as well.
</blockquote></details>

done
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 887 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-My8ftNb44vSb3E_2PvY:-My9cXYE542YKAPSQIM7:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/81efd64752251c5fe126c738c5b64c465ac9e600/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L887)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we write a test for this? Similar to `pkg/kv/kvserver/concurrency/testdata/concurrency_manager/wait_self`, which was added in https://github.com/cockroachdb/cockroach/commit/dc689ae9bdef495533c5aef2cc294485186201d3.
</blockquote></details>

done
___
*[pkg/kv/kvserver/concurrency/lock_table_waiter.go, line 908 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-My8k9k22YiSxeTP8hVw:-My94YvDCb74_InRNr_p:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d9ba98fceed0dd6f87cfaa31b5948cedfc03e73/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L908)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this need to be a pointer? I assume it causes the argument to `notify` to escape to the heap and raises (unfounded) aliasing concerns.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 7 files at r33, 9 of 9 files at r36, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-:-MyA0z-68ZTVRDFc5duh:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @andreimatei and @tbg)

___
*[pkg/kv/kvserver/concurrency/concurrency_manager_test.go, line 82 at r36](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MyA0rzECpS6gobZpYue:-MyA0rzECpS6gobZpYuf:bu41pa7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aa2f234c54fa561f4d489509b9076a5aaab8e13/pkg/kv/kvserver/concurrency/concurrency_manager_test.go#L82)):*
> ```Go
> // debug-set-clock           ts=<secs>
> // debug-set-discovered-locks-threshold-to-consult-finalized-txn-cache n=<count>
> // debug-set-max-locks <limit>
> ```

`n=<count>`


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-:-MyAFqHF0g39Q2dWNyUy:b-tlx7ko)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @nvanbenschoten and @tbg)

___
*[pkg/kv/kvserver/concurrency/concurrency_manager_test.go, line 82 at r36](https://reviewable.io/reviews/cockroachdb/cockroach/74878#-MyA0rzECpS6gobZpYue:-MyAFpLC0Z_o1-H1DNdO:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aa2f234c54fa561f4d489509b9076a5aaab8e13/pkg/kv/kvserver/concurrency/concurrency_manager_test.go#L82)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`n=<count>`
</blockquote></details>

done



_Code quote:_
```Go
// debug-set-max-locks <limit>
```


<!-- Sent from Reviewable.io -->
